internal/controller/util/json_util.go

```diff
@@ -34,38 +35,71 @@ func EvaluateCheckHook(client client.Client, hook *kubeobjects.HookSpec, log log
 		// handle pod type
 		resource := &corev1.Pod{}
 
-		err := WaitUntilResourceExists(client, resource, nsScopedName, time.Duration(timeout)*time.Second)
-		if err != nil {
-			return false, err
-		}
+		// err := WaitUntilResourceExists(client, resource, nsScopedName, time.Duration(timeout)*time.Second)
+		// if err != nil {
+		// 	return false, err
+		// }
 
-		return EvaluateCheckHookExp(hook.Chk.Condition, resource)
+		//return EvaluateCheckHookExp(hook.Chk.Condition, resource)
+		return EvaluateCheckHookExpWithTimeout(client, resource, nsScopedName, time.Duration(timeout)*time.Second, hook.Chk.Condition)
 	case "deployment":
 		// handle deployment type
 		resource := &appsv1.Deployment{}
 
-		err := WaitUntilResourceExists(client, resource, nsScopedName, time.Duration(timeout)*time.Second)
-		if err != nil {
-			return false, err
-		}
+		// err := WaitUntilResourceExists(client, resource, nsScopedName, time.Duration(timeout)*time.Second)
+		// if err != nil {
+		// 	return false, err
+		// }
 
-		return EvaluateCheckHookExp(hook.Chk.Condition, resource)
+		return EvaluateCheckHookExpWithTimeout(client, resource, nsScopedName, time.Duration(timeout)*time.Second, hook.Chk.Condition)
 	case "statefulset":
 		// handle statefulset type
 		resource := &appsv1.StatefulSet{}
 
-		err := WaitUntilResourceExists(client, resource, nsScopedName, time.Duration(timeout)*time.Second)
-		if err != nil {
-			return false, err
-		}
+		// err := WaitUntilResourceExists(client, resource, nsScopedName, time.Duration(timeout)*time.Second)
+		// if err != nil {
+		// 	return false, err
+		// }
 
-		return EvaluateCheckHookExp(hook.Chk.Condition, resource)
+		return EvaluateCheckHookExpWithTimeout(client, resource, nsScopedName, time.Duration(timeout)*time.Second, hook.Chk.Condition)
 	}
 
 	return false, nil
 }
 
-func WaitUntilResourceExists(client client.Client, obj client.Object, nsScopedName types.NamespacedName,
+func EvaluateCheckHookExpWithTimeout(client client.Client, obj client.Object, nsScopedName types.NamespacedName,
+	timeout time.Duration, booleanExpression string,
+) (bool, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	ticker := time.NewTicker(pollInterval * time.Millisecond) // Poll every 100 milliseconds
```

Comment:
This comment will quickly become stale when we change pollInternal. It is better to remove it.

However polling every 100 millisecond is too much. If we need to check a resource until the resource reach some state, the right way is to watch the resource. Each time the resources changes, you check the desired state.

Until we get a proper watch we can make this more efficient by using a backoff. We can start with short wait internal to be able to detect very short waits quickly (e.g. 10 ms), and double the interval on each retry, until we reach maximum internal (e.g. 10 seconds).